### DISCUSSION TITLE
Avoiding build timeouts

### DISCUSSION THREAD

**Discussion by jbelkins** [SwiftPackageIndex/SwiftPackageIndex-Server#2161] [Q&A] [answered]
Our [package](https://swiftpackageindex.com/awslabs/aws-sdk-swift) currently takes a long time to build because building the whole thing is a lot of code across ~320 package products.  All of our SPI builds that don't otherwise fail are currently timing out.

We totally get that SPI needs to limit package build time to conserve build minutes.  Accordingly, we'd like to customize our build so we can build a subset of our targets; i.e. enough to prove build compatibility.

I see the `.spi.yml` file allows a specific Xcode target or scheme to be specified as described [here](https://blog.swiftpackageindex.com/posts/the-swift-package-index-metadata-file-first-steps/); that will help reduce build time on our Apple platform builds.  Is there an equivalent mechanism for specifying the product on Linux builds performed with `swift build` instead of Xcode?

**Comment by finestructure** [comment, marked as answer]
Thanks for reporting this, @jbelkins, and great to hear that you're looking into working around our build time limit!

It's indeed been tricky for us to both allow for a range of different build times while also ensuring that we're not accumulating a backlog when "heavy" packages are publishing releases.

As you've seen, we support the `scheme:` key in `.spi.yml` to allow for package authors to specify an Xcode scheme for builds driven by Xcode (`macos-xcode`, `ios`, `watchos`, `tvos`).

There's also a `target:` key, which does a similar thing for SPM-based builds (`macos-spm`, `linux`). See the [SPIManifest documentation](https://swiftpackageindex.com/swiftpackageindex/spimanifest/0.13.0/documentation/spimanifest/commonusecases#Control-Targets-and-Schemes) for details.

So if that's an option, it might make sense to add a new target to you `Package.swift` file that builds a relevant subset and which you could refer to like so:

```yaml
version: 1
builder:
  configs:
  - platform: linux
    target: SPICompatibility
```

`SPICompatibility` is probably not an ideal name for this although I'm not sure what else to call it 🤔

**Reply by finestructure** [reply]
PS: You should be able to simplify that `spi.yml` to the following if you set up an Xcode scheme and an SPI target:

```yaml
version: 1
builder:
  configs:
    - scheme: SPICompatibility
      target: SPICompatibility
```

This will set `scheme` and `target` for all platforms and Xcode and SPM builds will pick their respective parameters and ignore the other.

**Reply by jbelkins** [reply]
Thanks for the help!  We're going to push up a `.spi.yml` file and see if we can address our issues.
